File: src/main/kotlin/viewmodel/TimeShareViewModel.kt
package viewmodel

import data.model.TimeSharePoint
import data.remote.getTimeShare
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.launch
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.util.logging.Logger

/**
 * 负责处理分时图业务逻辑和状态管理。
 * 从 UI 层分离出来，遵循 MVVM 模式。
 */
class TimeShareViewModel(private val code: String) {
    // 使用StateFlow来管理UI状态，确保线程安全和Compose的响应式更新
    private val _uiState = MutableStateFlow<TimeShareUiState>(TimeShareUiState.Loading)
    val uiState: StateFlow<TimeShareUiState> = _uiState

    // 日志记录器
    private val logger = Logger.getLogger("TimeShareViewModel")

    init {
        // ViewModel初始化时，立即在IO线程上加载数据
        loadTimeShareData()
    }

    private fun loadTimeShareData() {
        // 使用ViewModel的协程作用域，确保协程在ViewModel销毁时自动取消
        CoroutineScope(Dispatchers.IO).launch {
            try {
                // 调用API获取分时数据
                val response = getTimeShare(code)
                val dataContainer = response?.data
                if (dataContainer == null || dataContainer.trends.isEmpty()) {
                    _uiState.value = TimeShareUiState.NoData
                    logger.warning("股票[$code]的原始分时数据列表(trends)为空或数据容器(data)为null")
                } else {
                    val points = parseTrends(dataContainer.trends)
                    if (points.isEmpty()) {
                        _uiState.value = TimeShareUiState.NoData
                        logger.warning("解析后的股票[$code]分时数据点列表为空，可能所有数据点格式都有问题。")
                    } else {
                        _uiState.value = TimeShareUiState.Success(points, dataContainer.preClosePrice)
                        logger.info("成功加载并解析了股票[$code]的分时数据，共${points.size}个点，昨收价: ${dataContainer.preClosePrice}")
                    }
                }
            } catch (e: Exception) {
                _uiState.value = TimeShareUiState.Error("数据加载失败: ${e.message}")
                logger.severe("加载股票[$code]的分时数据时发生错误: ${e.message}")
            }
        }
    }

    private fun parseTrends(trends: List<String>): List<TimeSharePoint> {
        val formatter = DateTimeFormatter.ofPattern("HH:mm")
        return trends.mapNotNull { trendString ->
            val parts = trendString.split(',')
            try {
                val timeStr = parts[0].split(' ')[1]
                val time = LocalTime.parse(timeStr, formatter)
                val price = parts[1].toFloat()
                TimeSharePoint(time, price)
            } catch (e: Exception) {
                logger.warning("解析分时数据点失败: '$trendString', 错误: ${e.message}")
                null
            }
        }
    }
}

/**
 * 定义分时图UI状态的密封类。
 * 与 ViewModel 紧密相关，因此放在同一个文件中。
 */
sealed class TimeShareUiState {
    object Loading : TimeShareUiState()
    object NoData : TimeShareUiState()
    data class Success(val points: List<TimeSharePoint>, val preClosePrice: Float) : TimeShareUiState()
    data class Error(val message: String) : TimeShareUiState()
}

File: rvan/src/viewmodels/TimeShareViewModel.ts
import { makeAutoObservable, runInAction } from "mobx";
import { TimeSharePoint } from "../data/model/StockData";
import { fetchEastMoneyTimeShare } from "../data/remote/ApiClient";
import { listen } from "@tauri-apps/api/event";

/**
 * UI 状态类型
 */
export type TimeShareUiState =
  | { type: "loading" }
  | { type: "no-data" }
  | { type: "error"; message: string }
  | { type: "success"; points: TimeSharePoint[]; preClosePrice: number };

/**
 * 分时图 ViewModel
 */
export class TimeShareViewModel {
  private code: string;
  uiState: TimeShareUiState = { type: "loading" };

  constructor(code: string) {
    this.code = code;
    makeAutoObservable(this);
    this.loadTimeShareData();
    // 监听股票切换事件
    this.listenSwitchTargetEvent();
  }

  /**
   * 监听股票切换事件
   */
  private async listenSwitchTargetEvent(): Promise<void> {
    await listen<string>("switch-target-stock", (event) => {
      const newCode = event.payload;
      if (newCode !== this.code) {
        this.updateCode(newCode);
      }
    });
  }

  /**
   * 加载分时数据
   */
  async loadTimeShareData(): Promise<void> {
    try {
      const response = await fetchEastMoneyTimeShare(this.code);
      const json = JSON.parse(response);

      if (!json.data || !json.data.trends || json.data.trends.length === 0) {
        runInAction(() => {
          this.uiState = { type: "no-data" };
        });
        return;
      }

      const points = this.parseTrends(json.data.trends);
      const preClosePrice = json.data.preClose || 0;

      if (points.length === 0) {
        runInAction(() => {
          this.uiState = { type: "no-data" };
        });
      } else {
        runInAction(() => {
          this.uiState = {
            type: "success",
            points,
            preClosePrice,
          };
        });
      }
    } catch (error) {
      runInAction(() => {
        this.uiState = {
          type: "error",
          message: `数据加载失败：${error}`,
        };
      });
    }
  }

  /**
   * 解析分时数据
   */
  private parseTrends(trends: string[]): TimeSharePoint[] {
    return trends
      .map((trendString: string) => {
        const parts = trendString.split(",");
        try {
          const timeStr = parts[0].split(" ")[1];
          const price = parseFloat(parts[1]);
          return { time: timeStr, price };
        } catch (e) {
          console.warn("解析分时数据点失败:", trendString, e);
          return null;
        }
      })
      .filter((p: TimeSharePoint | null): p is TimeSharePoint => p !== null);
  }

  /**
   * 更新代码并重新加载
   */
  updateCode(code: string): void {
    if (this.code === code) {
      // 代码相同，只重新加载数据
      this.loadTimeShareData();
      return;
    }
    this.code = code;
    runInAction(() => {
      this.uiState = { type: "loading" };
    });
    this.loadTimeShareData();
  }
}
